src/jobservice/job/impl/gc/garbage_collection.go

```diff
@@ -169,60 +198,58 @@ func (gc *GarbageCollector) cleanCache() error {
-			return fmt.Errorf("error retrieving '%s' keys", pattern)
+// deleteCandidates deletes the two parts of artifact from harbor DB
+// 1, required part, the artifacts were removed from Harbor.
+// 2, optional part, the untagged artifacts.
+func (gc *GarbageCollector) deleteCandidates(ctx job.Context) error {
+	var untagged []*artifact.Artifact
+	// default is to clean trash
+	flushTrash := true
+	defer func(flushTrash bool) {
+		if flushTrash {
+			if err := gc.artrashMgr.Flush(ctx.SystemContext()); err != nil {
+				gc.logger.Errorf("failed to flush artifact trash")
+			}
 		}
-		iter, err = redis.Int(arr[0], nil)
-		if err != nil {
-			return fmt.Errorf("unexpected type for Int, got type %T", err)
-		}
-		k, err := redis.Strings(arr[1], nil)
-		if err != nil {
-			return fmt.Errorf("converts an array command reply to a []string %v", err)
-		}
-		keys = append(keys, k...)
+	}(flushTrash)
 
-		if iter == 0 {
-			break
-		}
+	// Mark
+	// get all of required and untagged artifacts
+	required, err := gc.artrashMgr.Filter(ctx.SystemContext())
+	if err != nil {
+		return nil
 	}
-	for _, key := range keys {
-		_, err := con.Do("DEL", key)
+	if gc.deleteUntagged {
+		untagged, err = gc.artCtl.List(ctx.SystemContext(), &q.Query{
+			Keywords: map[string]interface{}{
+				"Tags": "nil",
+			},
+		}, nil)
 		if err != nil {
-			return fmt.Errorf("failed to clean registry cache %v", err)
+			return err
 		}
 	}
-	return nil
-}
 
-func (gc *GarbageCollector) ensureQuota() error {
-	projects, err := dao.GetProjects(nil)
-	if err != nil {
-		return err
-	}
-	for _, project := range projects {
-		pSize, err := dao.CountSizeOfProject(project.ProjectID)
-		if err != nil {
-			gc.logger.Warningf("error happen on counting size of project:%d by artifact, error:%v, just skip it.", project.ProjectID, err)
-			continue
+	// Sweep
+	for _, art := range required {
+		if err := deleteManifest(gc.registryURL, art.RepositoryName, art.Digest); err != nil {
+			flushTrash = false
+			return fmt.Errorf("failed to delete manifest, %s:%s with error: %v", art.RepositoryName, art.Digest, err)
 		}
-		quotaMgr, err := common_quota.NewManager("project", strconv.FormatInt(project.ProjectID, 10))
-		if err != nil {
-			gc.logger.Errorf("Error occurred when to new quota manager %v, just skip it.", err)
-			continue
-		}
-		if err := quotaMgr.SetResourceUsage(types.ResourceStorage, pSize); err != nil {
-			gc.logger.Errorf("cannot ensure quota for the project: %d, err: %v, just skip it.", project.ProjectID, err)
-			continue
-		}
-		if err := dao.RemoveUntaggedBlobs(project.ProjectID); err != nil {
-			gc.logger.Errorf("cannot delete untagged blobs of project: %d, err: %v, just skip it.", project.ProjectID, err)
-			continue
+	}
+	if gc.deleteUntagged {
+		for _, art := range untagged {
+			repo, err := gc.repoMgr.Get(ctx.SystemContext(), art.RepositoryID)
+			if err != nil {
+				return err
+			}
+			if err := deleteManifest(gc.registryURL, repo.Name, art.Digest); err != nil {
```

Comment:
Only the manifest of the parent artifact is deleted, doesn't cover the children's.
I suggest that if the untagged artifacts are the candidates, delete them first, and this will move them(both parent and children) into the trash table, then you can  handle the trash

Reply:
updated